src/lib.rs

```diff
@@ -991,25 +992,90 @@ pub(crate) mod tests {
     }
 
     use proptest::prelude::*;
-    use proptest::sample::select;
+    use proptest::sample::{select, SizeRange};
+
+    #[derive(Clone, Debug, PartialEq, Eq)]
+    enum OpType {
+        Append,
+        Witness,
+        Unwitness,
+        Checkpoint,
+        Rewind,
+        Authpath,
+    }
+
+    fn arb_operations_0<H: Clone + Debug + 'static>(
+        optypes: Vec<OpType>,
+        items: Vec<H>,
+    ) -> impl Strategy<Value = Vec<Operation<H>>> {
+        let mut idx = 0;
+        optypes
+            .iter()
+            .map(move |t| {
+                let items = items.clone();
+                match t {
+                    OpType::Append => {
+                        let item = items.get(idx).unwrap().clone();
+                        idx += 1;
+                        Just(Append(item)).boxed()
+                    }
+                    OpType::Witness => Just(Witness).boxed(),
+                    OpType::Unwitness => (0..(idx + 1))
+                        .prop_map(move |i| {
+                            let item = items.get(i).unwrap().clone();
+                            Unwitness(Position::from(i), item)
+                        })
+                        .boxed(),
+                    OpType::Checkpoint => Just(Checkpoint).boxed(),
+                    OpType::Rewind => Just(Rewind).boxed(),
+                    OpType::Authpath => (0..(idx + 1))
+                        .prop_map(move |i| {
+                            let item = items.get(i).unwrap().clone();
+                            Authpath(Position::from(i), item)
+                        })
+                        .boxed(),
+                }
+            })
+            .collect::<Vec<_>>()
+    }
 
-    fn arb_operation<G: Strategy>(item_gen: G) -> impl Strategy<Value = Operation<G::Value>>
+    fn arb_operations_1<G: Strategy>(
+        item_gen: G,
+        optypes: Vec<OpType>,
+    ) -> impl Strategy<Value = Vec<Operation<G::Value>>>
     where
         G::Value: Clone + 'static,
     {
-        item_gen.prop_flat_map(|item| {
+        let append_count = optypes
+            .iter()
+            .filter(|t| matches!(t, OpType::Append))
+            .count();
+        proptest::collection::vec(item_gen, append_count + 2)
```

Comment:
Why is this `+ 2`?

Reply:
That should be + 1, I must have fat-fingered it. We want one more item than the number of appends so that we can always generate an `Authpath` or `Unwitness` even if we don't append any notes to the tree (those would always return false/None from the relevant operations).
```suggestion
        proptest::collection::vec(item_gen, append_count + 1)
```